crates/hir_ty/src/diagnostics/expr.rs

```diff
@@ -141,6 +145,30 @@ impl ExprValidator {
             });
     }
 
+    fn check_for_trailing_return(&mut self, db: &dyn HirDatabase) {
+        // todo: handle inner functions and lambdas.
```

Comment:
Inner functions I believe should already be covered by this as they are their own `DefWithBodyId` again.

For lambdas you will just have to go iterate through all expressions of the body and search for a lambda expression, then do the logic you already do here for it's body expression again